### ISSUE TITLE
Flux.single() deadlock with Flux.create() and Mono.block()

### ISSUE THREAD

**Issue by cnaj** [reactor/reactor-core#3621]
## Expected Behavior
* Reactor should behave the same, regardless of version used
* Mono.block() should produce a result

## Actual Behavior
* Testcase behavior differs between versions
* Mono.block() blocks for an indefinite time (3.4.32, 3.4.33).

## Steps to Reproduce
`testDeadlock()` (see below) runs into timeout with reactor-core version 3.4.32 and 3.4.33, but is successful with 3.4.31 and 3.5.11

```java
    @Test
    void testWorks() {
        Flux<String> flux = Flux.create(sink -> {
            sink.next("hello");
            sink.complete();
        });

        Mono<String> mono = flux
                .filter("hello"::equals)
                .single();

        String result = mono.block();
        assertThat(result)
                .isEqualTo("hello");
    }

    @Test
    @Timeout(value = 10)
    void testDeadlock() {
        Flux<String> flux = Flux.create(sink -> {
            sink.next("hello");
            sink.next("world");
            sink.complete();
        });

        Mono<String> mono = flux
                .filter("hello"::equals)
                .single();

        String result = mono.block();
        assertThat(result)
                .isEqualTo("hello");
    }
```

## Your Environment

* Reactor version(s) used: 3.4.31, 3.4.32, 3.4.33, 3.5.11
* JVM version (`java -version`): openjdk version "17.0.8.1"
* OS and version (eg `uname -a`): Ubuntu 22.04.3 LTS

**Comment by cnaj** [comment]
Edited initial bugreport with additional reactor-core versions.

This behavior appears in 3.4.x versions starting from 3.4.32 (which are included in Spring Boot versions 2.7.15 upwards). The 3.5.x versions seem not to be affected (included in Spring Boot 3.x).

Note: Upgrading Spring Boot from 2.7.x to 3.x is not easily possible for everyone because Java 17 is required for Spring Boot 3.x.

**Comment by OlegDokuka** [comment]
this is related to #3569 which for any reasons was not back ported to 3.4.x 🫣.

**Comment by OlegDokuka** [comment]
should be fixed with https://github.com/reactor/reactor-core/commit/0b988cb1cef0339df81220b937e75fff5c7c9123

**Comment by cnaj** [comment]
For reference: Spring Boot 2.7.18 contains the fixed Reactor version 3.4.34.
